In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\Josh Abrams\Desktop\09-2022 Instacart Basket Analysis'

In [3]:
df_prods = pd.read_csv(r'C:\Users\Josh Abrams\Desktop\09-2022 Instacart Basket Analysis\Data\Original Data\products.csv', index_col = False)

In [4]:
df_ords = pd.read_csv(r'C:\Users\Josh Abrams\Desktop\09-2022 Instacart Basket Analysis\Data\Prepared Data\orders_wrangled.csv', index_col = False)

In [5]:
df_ords.describe()



Unnamed: 0      order_id   customer_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [6]:
df_ords

Unnamed: 0  order_id  customer_id eval_set  order_number  \
0                 0   2539329            1    prior             1   
1                 1   2398795            1    prior             2   
2                 2    473747            1    prior             3   
3                 3   2254736            1    prior             4   
4                 4    431534            1    prior             5   
...             ...       ...          ...      ...           ...   
3421078     3421078   2266710       206209    prior            10   
3421079     3421079   1854736       206209    prior            11   
3421080     3421080    626363       206209    prior            12   
3421081     3421081   2977660       206209    prior            13   
3421082     3421082    272231       206209    train            14   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
1                         3                  7                    15.0  
2                         3                 12                    21.0  
3                         4                  7                    29.0  
4                         4                 15                    28.0  
...                     ...                ...                     ...  
3421078                   5                 18                    29.0  
3421079                   4                 10                    30.0  
3421080                   1                 12                    18.0  
3421081                   1                 12                     7.0  
3421082                   6                 14                    30.0  

[3421083 rows x 8 columns]

In [7]:
df_ords = df_ords.drop(columns = ['Unnamed: 0']) 

In [8]:
df_ords

order_id  customer_id eval_set  order_number  orders_day_of_week  \
0         2539329            1    prior             1                   2   
1         2398795            1    prior             2                   3   
2          473747            1    prior             3                   3   
3         2254736            1    prior             4                   4   
4          431534            1    prior             5                   4   
...           ...          ...      ...           ...                 ...   
3421078   2266710       206209    prior            10                   5   
3421079   1854736       206209    prior            11                   4   
3421080    626363       206209    prior            12                   1   
3421081   2977660       206209    prior            13                   1   
3421082    272231       206209    train            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [9]:
df_ords.describe()


order_id   customer_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [10]:
# create a dataframe
df_test= pd.DataFrame()

In [11]:
# Create a mixed type column
df_test['mix'] = ['a','b',1, True]

In [12]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [13]:
#The function for checking whether a dataframe contains any mixed-type columns is as follows
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [14]:
df_test['mix'] = df_test['mix'].astype('str')

In [15]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [16]:
df_nan = df_prods [df_prods ['product_name'].isnull() == True]

In [17]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

## Addressing Missing Values


In [18]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [19]:
df_prods.shape

(49693, 5)

In [20]:
df_prods_clean=df_nan = df_prods [df_prods ['product_name'].isnull() == False]

In [21]:
df_prods_clean.shape


(49677, 5)

In [22]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [23]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [24]:
df_prods_clean.shape

(49677, 5)

In [25]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [26]:
df_prods_clean_no_dups.shape

(49672, 5)

In [27]:
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Data','Prepared Data', 'products_checked.csv'))


## TASK 4-5

## 2. Run the df.describe() function on your df_prods dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further. Tip: Keep an eye on min and max values!

In [28]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

## All things are consistent and expected except for the max price.

## 3.Check for mixed-type data in your df_ords dataframe.

In [29]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

## 4.If you find mixed-type data, fix it. The column in question should contain observations of a single data type

In [30]:
# Task 3 - If you find a mixed-type data, fix it. 
# No mixed data was reported. 

## 5.Run a check for missing values in your df_ords dataframe. In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [31]:
df_ords.isnull().sum()

order_id                       0
customer_id                    0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [32]:
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [33]:
df_ords_nan

order_id  customer_id eval_set  order_number  orders_day_of_week  \
0         2539329            1    prior             1                   2   
11        2168274            2    prior             1                   2   
26        1374495            3    prior             1                   1   
39        3343014            4    prior             1                   6   
45        2717275            5    prior             1                   3   
...           ...          ...      ...           ...                 ...   
3420930    969311       206205    prior             1                   4   
3420934   3189322       206206    prior             1                   3   
3421002   2166133       206207    prior             1                   6   
3421019   2227043       206208    prior             1                   1   
3421069   3154581       206209    prior             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

##  
 - There are 206209 missing values in the 'days_since_prior_order' column.
 - The descriptive stats of the 'days_since_prior_order' are unusal and erratic. 
 - There are way too many missing values to remove them from the analysis
 - The most likely reason for this many missing values in this column is because they are new customers, therefore, do not have previous orders have signed up but have never placed an order. 
 - I would create a 'first_time_customer' flag and tie it to any of these missing values so that we can keep the rest of the complete and valuable data. `
    

## 6.Address the missing values using an appropriate method. In a markdown cell, explain why you used your method of choice.


In [34]:
df_ords_clean = df_ords

In [35]:
df_ords_clean['new_customer'] = df_ords['days_since_prior_order'].isnull() == True

In [36]:
df_ords_clean.head()

order_id  customer_id eval_set  order_number  orders_day_of_week  \
0   2539329            1    prior             1                   2   
1   2398795            1    prior             2                   3   
2    473747            1    prior             3                   3   
3   2254736            1    prior             4                   4   
4    431534            1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  new_customer  
0                  8                     NaN          True  
1                  7                    15.0         False  
2                 12                    21.0         False  
3                  7                    29.0         False  
4                 15                    28.0         False

## 7. Run a check for duplicate values in your df_ords data. In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [37]:
df_ords_clean[df_ords.duplicated()]

Empty DataFrame
Columns: [order_id, customer_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, new_customer]
Index: []

## 8.Address the duplicates using an appropriate method. In a markdown cell, explain why you used your method of choice.

In [38]:
## There are no duplicates in df.ords_clean

## 9.Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [39]:
df_ords_clean.to_csv(os.path.join(path, 'Data', 'Cleaned Data', 'orders_clean.csv'))

In [40]:
df_prods_clean.to_csv(os.path.join(path, 'Data', 'Cleaned Data', 'products_clean.csv'))